### ISSUE TITLE
Dart2JS crashes when compiling element

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#31803]
```
Internal Error: The compiler crashed when compiling this element.
  const factory ExistingProvider.forToken(
    OpaqueToken<T> token,
    Object useExisting, {
    bool multi,
  }) = ExistingProvider<T>._;

The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (2.0.0-dev.13.0), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

The compiler crashed: NoSuchMethodError: The getter 'maxDepth' was called on null.
Receiver: null
Tried calling: maxDepth
#0      Object.noSuchMethod (dart:core-patch/dart:core/object_patch.dart:46)
#1      ElementX&AstElementMixin&AnalyzableElementX&ClassElementCommon.hierarchyDepth (package:compiler/src/elements/common.dart:198:50)
#2      ElementX&AstElementMixin&AnalyzableElementX&ClassElementCommon.asInstanceOf (package:compiler/src/elements/common.dart:203:55)
#3      ResolutionInterfaceType.asInstanceOf (package:compiler/src/elements/resolution_types.dart:532:49)
#4      SubtypeVisitor&AbstractTypeRelationMixin.asInstanceOf (package:compiler/src/elements/resolution_types.dart:1067:12)
#5      AbstractTypeRelation.visitInterfaceType (package:compiler/src/elements/types.dart:858:32)
#6      ResolutionInterfaceType.accept (package:compiler/src/elements/resolution_types.dart:562:20)
#7      MoreSpecificVisitor.isMoreSpecific (package:compiler/src/elements/types.dart:1043:14)
#8      SubtypeVisitor.isSubtype (package:compiler/src/elements/types.dart:1072:32)
#9      SubtypeVisitor.isAssignable (package:compiler/src/elements/types.dart:1076:12)
#10     SubtypeVisitor.invalidFunctionParameterTypes (package:compiler/src/elements/types.dart:1088:13)
#11     AbstractTypeRelation.visitFunctionTypeInternal.incompatibleParameters (package:compiler/src/elements/types.dart:922:13)
#12     AbstractTypeRelation.visitFunctionTypeInternal (package:compiler/src/elements/types.dart:931:31)
#13     AbstractTypeRelation.visitFunctionType (package:compiler/src/elements/types.dart:897:19)
#14     ResolutionFunctionType.accept (package:compiler/src/elements/resolution_types.dart:777:20)
#15     MoreSpecificVisitor.isMoreSpecific (package:compiler/src/elements/types.dart:1043:14)
#16     SubtypeVisitor.isSubtype (package:compiler/src/elements/types.dart:1072:32)
#17     Types.isSubtype (package:compiler/src/elements/resolution_types.dart:1219:27)
#18     ResolverVisitor.visitRedirectingFactoryBody (package:compiler/src/resolution/members.dart:3688:26)
#19     RedirectingFactoryBody.accept (package:compiler/src/tree/nodes.dart:1668:38)
#20     CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#21     ResolverTask.resolveMethodElementImplementation.<anonymous closure> (package:compiler/src/resolution/resolution.dart:268:17)
#22     CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#23     ResolverTask.resolveMethodElementImplementation (package:compiler/src/resolution/resolution.dart:225:21)
#24     ResolverTask.resolveMethodElement.<anonymous closure> (package:compiler/src/resolution/resolution.dart:358:16)
#25     CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#26     ResolverTask.resolveMethodElement (package:compiler/src/resolution/resolution.dart:310:21)
#27     ResolverTask.resolve.<anonymous closure> (package:compiler/src/resolution/resolution.dart:106:32)
#28     CompilerTask.measure (package:compiler/src/common/tasks.dart:63:44)
#29     ResolverTask.resolve (package:compiler/src/resolution/resolution.dart:84:12)
#30     CompilerResolution.computeWorldImpact.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:1520:64)
#31     _HashVMBase&MapMixin&&_LinkedHashMapMixin.putIfAbsent (dart:collection-patch/dart:collection/compact_hash.dart:275)
#32     CompilerResolution.computeWorldImpact.<anonymous closure> (package:compiler/src/compiler.dart:1516:32)
#33     CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#34     CompilerResolution.computeWorldImpact (package:compiler/src/compiler.dart:1498:31)
#35     CompilerResolution.ensureResolved (package:compiler/src/compiler.dart:1418:5)
#36     CompileTimeConstantEvaluator.createConstructorInvocation (package:compiler/src/compile_time_constants.dart:915:16)
#37     CompileTimeConstantEvaluator.visitNewExpression (package:compiler/src/compile_time_constants.dart:901:12)
#38     NewExpression.accept (package:compiler/src/tree/nodes.dart:872:38)
#39     CompileTimeConstantEvaluator.evaluate (package:compiler/src/compile_time_constants.dart:416:31)
#40     ConstantCompilerBase.compileNodeWithDefinitions (package:compiler/src/compile_time_constants.dart:320:38)
#41     DartConstantCompiler.compileNodeWithDefinitions (package:compiler/src/compile_time_constants.dart:386:41)
#42     ConstantCompilerBase.compileVariableWithDefinitions (package:compiler/src/compile_time_constants.dart:255:20)
#43     ConstantCompilerBase.internalCompileVariable.<anonymous closure> (package:compiler/src/compile_time_constants.dart:219:37)
#44     CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#45     ConstantCompilerBase.internalCompileVariable (package:compiler/src/compile_time_constants.dart:215:21)
#46     ConstantCompilerBase.compileConstant (package:compiler/src/compile_time_constants.dart:176:12)
#47     CompileTimeConstantEvaluator.visitSend (package:compiler/src/compile_time_constants.dart:646:39)
#48     Send.accept (package:compiler/src/tree/nodes.dart:716:38)
#49     CompileTimeConstantEvaluator.evaluateConstant (package:compiler/src/compile_time_constants.dart:425:31)
#50     CompileTimeConstantEvaluator.visitLiteralList (package:compiler/src/compile_time_constants.dart:466:30)
#51     LiteralList.accept (package:compiler/src/tree/nodes.dart:1454:38)
#52     CompileTimeConstantEvaluator.evaluateConstant (package:compiler/src/compile_time_constants.dart:425:31)
#53     CompileTimeConstantEvaluator.createConstructorInvocation (package:compiler/src/compile_time_constants.dart:957:41)
#54     CompileTimeConstantEvaluator.visitNewExpression (package:compiler/src/compile_time_constants.dart:901:12)
#55     NewExpression.accept (package:compiler/src/tree/nodes.dart:872:38)
#56     CompileTimeConstantEvaluator.evaluate (package:compiler/src/compile_time_constants.dart:416:31)
#57     ConstantCompilerBase.compileNodeWithDefinitions (package:compiler/src/compile_time_constants.dart:320:38)
#58     DartConstantCompiler.compileNodeWithDefinitions (package:compiler/src/compile_time_constants.dart:386:41)
#59     ConstantCompilerBase.compileMetadata (package:compiler/src/compile_time_constants.dart:364:12)
#60     JavaScriptConstantTask.compileMetadata.<anonymous closure> (package:compiler/src/js_backend/constant_handler_javascript.dart:93:32)
#61     CompilerTask.measure (package:compiler/src/common/tasks.dart:63:44)
#62     JavaScriptConstantTask.compileMetadata (package:compiler/src/js_backend/constant_handler_javascript.dart:91:12)
#63     ResolverTask.resolveMetadataAnnotation.<anonymous closure>.<anonymous closure> (package:compiler/src/resolution/resolution.dart:1106:62)
#64     CompilerTask.measure (package:compiler/src/common/tasks.dart:63:44)
#65     ResolverTask.resolveMetadataAnnotation.<anonymous closure> (package:compiler/src/resolution/resolution.dart:1082:15)
#66     CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#67     ResolverTask.resolveMetadataAnnotation (package:compiler/src/resolution/resolution.dart:1080:14)
#68     CompilerResolution.resolveMetadataAnnotation (package:compiler/src/compiler.dart:1399:24)
#69     MetadataAnnotationX.ensureResolved (package:compiler/src/elements/modelx.dart:3543:18)
#70     ResolverTask._postProcessClassElement (package:compiler/src/resolution/resolution.dart:698:16)
#71     ResolverTask._resolveTypeDeclaration.<anonymous closure>.<anonymous closure> (package:compiler/src/resolution/resolution.dart:613:15)
#72     CompilerTask.measure (package:compiler/src/common/tasks.dart:63:44)
#73     ResolverTask._resolveTypeDeclaration.<anonymous closure> (package:compiler/src/resolution/resolution.dart:600:14)
#74     CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#75     ResolverTask._resolveTypeDeclaration (package:compiler/src/resolution/resolution.dart:599:21)
#76     ResolverTask.resolveClass (package:compiler/src/resolution/resolution.dart:638:12)
#77     CompilerResolution.resolveClass (package:compiler/src/compiler.dart:1389:24)
#78     BaseClassElementX.ensureResolved (package:compiler/src/elements/modelx.dart:2889:18)
#79     ResolverVisitor.handleClassSend (package:compiler/src/resolution/members.dart:2128:9)
#80     ResolverVisitor.handleResolvedSend (package:compiler/src/resolution/members.dart:2898:14)
#81     ResolverVisitor.handleUnqualifiedSend (package:compiler/src/resolution/members.dart:2987:14)
#82     ResolverVisitor.visitSend (package:compiler/src/resolution/members.dart:3031:14)
#83     Send.accept (package:compiler/src/tree/nodes.dart:716:38)
#84     CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#85     ResolverVisitor.resolveArguments (package:compiler/src/resolution/members.dart:711:33)
#86     ResolverVisitor.handleStaticOrTopLevelAccess (package:compiler/src/resolution/members.dart:2658:41)
#87     ResolverVisitor.handleResolvedSend (package:compiler/src/resolution/members.dart:2909:14)
#88     ResolverVisitor.handleLibraryPrefixSend (package:compiler/src/resolution/members.dart:2173:16)
#89     ResolverVisitor.handlePrefixSend (package:compiler/src/resolution/members.dart:2247:16)
#90     ResolverVisitor.handleQualifiedSend (package:compiler/src/resolution/members.dart:2365:14)
#91     ResolverVisitor.visitSend (package:compiler/src/resolution/members.dart:3028:14)
#92     Send.accept (package:compiler/src/tree/nodes.dart:716:38)
#93     CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#94     ResolverVisitor.visitReturn (package:compiler/src/resolution/members.dart:3601:5)
#95     Return.accept (package:compiler/src/tree/nodes.dart:1609:38)
#96     CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#97     ResolverVisitor.visitFunctionExpression (package:compiler/src/resolution/members.dart:605:5)
#98     FunctionExpression.accept (package:compiler/src/tree/nodes.dart:1209:38)
#99     CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#100    ResolverVisitor.resolveArguments (package:compiler/src/resolution/members.dart:711:33)
#101    ResolverVisitor.handleDynamicAccessSemantics (package:compiler/src/resolution/members.dart:2287:11)
#102    ResolverVisitor.handleQualifiedSend (package:compiler/src/resolution/members.dart:2375:14)
#103    ResolverVisitor.visitSend (package:compiler/src/resolution/members.dart:3028:14)
#104    Send.accept (package:compiler/src/tree/nodes.dart:716:38)
#105    CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#106    ResolverVisitor.visitExpressionStatement (package:compiler/src/resolution/members.dart:538:5)
#107    ExpressionStatement.accept (package:compiler/src/tree/nodes.dart:1695:38)
#108    CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#109    ResolverVisitor.visitNodeList (package:compiler/src/resolution/members.dart:3575:7)
#110    NodeList.accept (package:compiler/src/tree/nodes.dart:924:38)
#111    CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#112    ResolverVisitor.visitIn (package:compiler/src/resolution/members.dart:501:31)
#113    ResolverVisitor.visitBlock (package:compiler/src/resolution/members.dart:521:5)
#114    Block.accept (package:compiler/src/tree/nodes.dart:980:38)
#115    CommonResolverVisitor.visit (package:compiler/src/resolution/resolution_common.dart:32:54)
#116    ResolverTask.resolveMethodElementImplementation.<anonymous closure> (package:compiler/src/resolution/resolution.dart:268:17)
#117    CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#118    ResolverTask.resolveMethodElementImplementation (package:compiler/src/resolution/resolution.dart:225:21)
#119    ResolverTask.resolveMethodElement.<anonymous closure> (package:compiler/src/resolution/resolution.dart:358:16)
#120    CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#121    ResolverTask.resolveMethodElement (package:compiler/src/resolution/resolution.dart:310:21)
#122    ResolverTask.resolve.<anonymous closure> (package:compiler/src/resolution/resolution.dart:106:32)
#123    CompilerTask.measure (package:compiler/src/common/tasks.dart:63:44)
#124    ResolverTask.resolve (package:compiler/src/resolution/resolution.dart:84:12)
#125    CompilerResolution.computeWorldImpact.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:1520:64)
#126    _HashVMBase&MapMixin&&_LinkedHashMapMixin.putIfAbsent (dart:collection-patch/dart:collection/compact_hash.dart:275)
#127    CompilerResolution.computeWorldImpact.<anonymous closure> (package:compiler/src/compiler.dart:1516:32)
#128    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#129    CompilerResolution.computeWorldImpact (package:compiler/src/compiler.dart:1498:31)
#130    _ResolutionWorkItem.run (package:compiler/src/common/resolution.dart:60:37)
#131    Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:774:70)
#132    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#133    Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:774:32)
#134    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#135    Compiler.emptyQueue.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:772:28)
#136    CompilerDiagnosticReporter.withCurrentElement (package:compiler/src/compiler.dart:1145:15)
#137    Compiler.emptyQueue.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:770:18)
#138    EnqueuerStrategy.processWorkItem (package:compiler/src/enqueue.dart:508:6)
#139    ResolutionEnqueuer.forEach (package:compiler/src/enqueue.dart:390:20)
#140    Compiler.emptyQueue.<anonymous closure> (package:compiler/src/compiler.dart:766:16)
#141    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#142    Compiler.emptyQueue (package:compiler/src/compiler.dart:765:14)
#143    Compiler.processQueue.<anonymous closure> (package:compiler/src/compiler.dart:787:7)
#144    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#145    Compiler.processQueue (package:compiler/src/compiler.dart:784:14)
#146    Compiler.compileLoadedLibraries.<anonymous closure> (package:compiler/src/compiler.dart:583:9)
#147    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#148    Compiler.compileLoadedLibraries (package:compiler/src/compiler.dart:529:16)
#149    Compiler.runInternal (package:compiler/src/compiler.dart:483:5)
<asynchronous suspension>
#150    Compiler.run.<anonymous closure>.<anonymous closure> (package:compiler/src/compiler.dart:313:38)
#151    new Future.sync (dart:async/future.dart:222)
#152    Compiler.run.<anonymous closure> (package:compiler/src/compiler.dart:313:20)
#153    CompilerTask.measureSubtask (package:compiler/src/common/tasks.dart:178:35)
#154    Compiler.run (package:compiler/src/compiler.dart:310:41)
#155    CompilerImpl.run.<anonymous closure>.<anonymous closure> (package:compiler/src/apiimpl.dart:265:25)
#156    _RootZone.runUnary (dart:async/zone.dart:1381)
#157    _FutureListener.handleValue (dart:async/future_impl.dart:129)
#158    _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:633)
#159    _Future._propagateToListeners (dart:async/future_impl.dart:662)
#160    _Future._complete (dart:async/future_impl.dart:467)
#161    _SyncCompleter.complete (dart:async/future_impl.dart:51)
#162    _completeOnAsyncReturn (dart:async-patch/dart:async/async_patch.dart:255)
#163    BazelInputProvider.readFromUri (package:compiler/src/source_file_provider.dart:509:5)
#164    _RootZone.runUnary (dart:async/zone.dart:1381)
#165    _FutureListener.handleValue (dart:async/future_impl.dart:129)
#166    _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:633)
#167    _Future._propagateToListeners (dart:async/future_impl.dart:662)
#168    _Future._completeWithValue (dart:async/future_impl.dart:477)
#169    _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:507)
#170    _microtaskLoop (dart:async/schedule_microtask.dart:41)
#171    _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
#172    _runPendingImmediateCallback (dart:isolate-patch/dart:isolate/isolate_patch.dart:111)
#173    _RawReceivePortImpl._handleMessage (dart:isolate-patch/dart:isolate/isolate_patch.dart:164)
```

**Comment by matanlurey** [comment]
I might have a workaround. If I don't, this will be a P1/P0.

**Comment by matanlurey** [comment]
Nope, no workaround. I can try and get a more minimal reproduction.

**Comment by matanlurey** [comment]
@hterkelsen mentioned he's fixed this stack trace before, but we might not have a new enough SDK in (/cc @keertip). If not, we definitely need this fixed as soon as possible (this week ideally).

**Comment by hterkelsen** [comment]
Fixed with 3c77ef673831e99868079d280dfd6e776a694fc1

**Comment by hterkelsen** [comment]
It turns out that 3c77ef673831e99868079d280dfd6e776a694fc1 does not really fix the issue. I'm still looking into it
